hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java

```diff
@@ -347,22 +346,20 @@ public boolean requiresDirtyCheck(Object entity) {
 	@SuppressWarnings( {"SimplifiableIfStatement"})
 	private boolean isUnequivocallyNonDirty(Object entity) {
 		if ( entity instanceof SelfDirtinessTracker ) {
-			boolean uninitializedProxy = false;
+			boolean mayBeDirty = true;
 			if ( entity instanceof PersistentAttributeInterceptable ) {
 				final PersistentAttributeInterceptable interceptable = (PersistentAttributeInterceptable) entity;
 				final PersistentAttributeInterceptor interceptor = interceptable.$$_hibernate_getInterceptor();
 				if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
-					EnhancementAsProxyLazinessInterceptor enhancementAsProxyLazinessInterceptor = (EnhancementAsProxyLazinessInterceptor) interceptor;
 					// When a proxy has dirty attributes, we have to treat it like a normal entity to flush changes
-					uninitializedProxy = !enhancementAsProxyLazinessInterceptor.isInitialized() && !( (SelfDirtinessTracker) entity ).$$_hibernate_hasDirtyAttributes();
+					mayBeDirty = ((EnhancementAsProxyLazinessInterceptor) interceptor).mayBeDirty();
```

Comment:
I think we might consider more entities to be dirty due to this change, possibly leading to other performance problems. What is the reason for not considering the `SelfDirtinessTracker#$$_hibernate_hasDirtyAttributes()` method if the `enhancementAsProxyLazinessInterceptor.isInitialized()` is `true`?

Reply:
the issue is that `( (SelfDirtinessTracker) entity ).$$_hibernate_hasDirtyAttributes()` with an entity having a collection initialized ad the field level like the following
```
class User {
  public Set<Role> roles = new HashSet<>();
}
```
returns always true.
By the way I think I have a little improvement for this part.

Reply:
But when is this relevant? If the entity is initialized and a hash set is assigned, then we can't know if it is dirty and thus have to consider it "maybe dirty". If the entity is not initialized, this is irrelevant and "maybe dirty" should be false.

Reply:
the problem is that in the specific case the entity is not initialized but we return that it may be dirty

Reply:
How is that it "is may be dirty" if it's not initialized ?

Reply:
if the entity is annotated with `@DYnamicUpdate` when we set a field the `EnhancementAsProxyLazinessInterceptor` does not initialize the entity but it keep track of the attribute name in  the `writtenFieldNames` set.